From d3c3e4f435e0f7ca396ae0a210dfaf261765ab34 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sun, 30 Nov 2025 18:40:58 +0100
Subject: [PATCH] gluon-core: use pubkey_privacy from provider

This fixes the pubkey_privacy not being in uci at all.
If the provider has a pubkey_privacy function, this can be used to override the existing behavior to look in
 site.mesh_vpn.pubkey_privacy ~= false
---
 .../luasrc/usr/lib/lua/gluon/mesh-vpn.lua                   | 6 ++++--
 .../luasrc/usr/lib/lua/gluon/mesh-vpn/provider/fastd.lua    | 4 ++++
 .../usr/lib/lua/gluon/mesh-vpn/provider/wireguard.lua       | 5 +++++
 3 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/package/gluon-mesh-vpn-core/luasrc/usr/lib/lua/gluon/mesh-vpn.lua b/package/gluon-mesh-vpn-core/luasrc/usr/lib/lua/gluon/mesh-vpn.lua
index 6547727aff..9f0de57ee0 100644
--- a/package/gluon-mesh-vpn-core/luasrc/usr/lib/lua/gluon/mesh-vpn.lua
+++ b/package/gluon-mesh-vpn-core/luasrc/usr/lib/lua/gluon/mesh-vpn.lua
@@ -45,7 +45,7 @@ function M.get_active_provider()
 end
 
 function M.get_enabled_public_key()
-	if M.enabled() ~= true or uci:get_bool('gluon', 'mesh_vpn', 'pubkey_privacy') ~= true then
+	if M.enabled() ~= true then
 		return nil
 	end
 
@@ -53,7 +53,9 @@ function M.get_enabled_public_key()
 
 	local pubkey
 	if active_vpn ~= nil then
-		pubkey = active_vpn.public_key()
+		if not active_vpn.pubkey_privacy() then
+			pubkey = active_vpn.public_key()
+		end
 	end
 
 	return pubkey
diff --git a/package/gluon-mesh-vpn-fastd/luasrc/usr/lib/lua/gluon/mesh-vpn/provider/fastd.lua b/package/gluon-mesh-vpn-fastd/luasrc/usr/lib/lua/gluon/mesh-vpn/provider/fastd.lua
index c6be84c4d6..5b84997d36 100644
--- a/package/gluon-mesh-vpn-fastd/luasrc/usr/lib/lua/gluon/mesh-vpn/provider/fastd.lua
+++ b/package/gluon-mesh-vpn-fastd/luasrc/usr/lib/lua/gluon/mesh-vpn/provider/fastd.lua
@@ -73,4 +73,8 @@ function M.mtu()
 	return site.mesh_vpn.fastd.mtu()
 end
 
+function M.pubkey_privacy()
+	return site.mesh_vpn.pubkey_privacy(true)
+end
+
 return M
diff --git a/package/gluon-mesh-vpn-wireguard/luasrc/usr/lib/lua/gluon/mesh-vpn/provider/wireguard.lua b/package/gluon-mesh-vpn-wireguard/luasrc/usr/lib/lua/gluon/mesh-vpn/provider/wireguard.lua
index 9837a8696e..ec5caad2c7 100644
--- a/package/gluon-mesh-vpn-wireguard/luasrc/usr/lib/lua/gluon/mesh-vpn/provider/wireguard.lua
+++ b/package/gluon-mesh-vpn-wireguard/luasrc/usr/lib/lua/gluon/mesh-vpn/provider/wireguard.lua
@@ -43,4 +43,9 @@ function M.mtu()
 	return site.mesh_vpn.wireguard.mtu()
 end
 
+function M.pubkey_privacy()
+	-- WireGuard inherently doesn't leak the pubkey to third parties upon initiating connections.
+	return false
+end
+
 return M
